A new command to read types' description from Swagger - kubectl explain

andronat opened the pull request:
This is a simple kubectl tool that parses the swagger schema and returns useful information for the various resources and their fields that we have.

Here you can see an example output:

```
# kubectl explain pods
DESCRIPTION:
Pod is a collection of containers that can run on a host. This resource is created by clients and scheduled onto hosts.

FIELDS:
   kind <string>
     Kind is a string value representing the REST resource this object
     represents. Servers may infer this from the endpoint the client submits
     requests to. Cannot be updated. In CamelCase. More info:
     http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#types-kinds

   metadata <Object>
     Standard object's metadata. More info:
     http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata

   spec <Object>
     Specification of the desired behavior of the pod. More info:
     http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status

   status   <Object>
     Most recently observed status of the pod. This data may not be up to date.
     Populated by the system. Read-only. More info:
     http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status

   apiVersion   <string>
     APIVersion defines the versioned schema of this representation of an
     object. Servers should convert recognized schemas to the latest internal
     value, and may reject unrecognized values. More info:
     http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#resources
```

and has this interface:

```
# kubectl help explain
Documentation of resources.

Possible resource types include: pods (po), services (svc),
replicationcontrollers (rc), nodes (no), events (ev), componentstatuses (cs),
limitranges (limits), persistentvolumes (pv), persistentvolumeclaims (pvc),
resourcequotas (quota), namespaces (ns) or endpoints (ep).

Usage:
  kubectl explain RESOURCE [flags]

Examples:
# Get the documentation of the resource and its fields
$ kubectl explain pods

# Get the documentation of a specific field of a resource
$ kubectl explain pods.spec.containers

Flags:
      --recursive[=false]: Print the fields of fields (Currently only 1 level deep)
```

/cc @ghodss

andronat committed:
client can return swagger schema

andronat committed:
new kubectl explain command

andronat commented:
Lots of changes. Added a dot notation feature for better exploration. Thanks @ghodss for the nice ideas and suggestions :)

bgrant0607 commented:
cc @kubernetes/kubectl

ghodss commented:
This implements #3060.

deads2k commented:
@pweil- This ought to help the detection issue you wrote in openshift.

jwforres commented:
cc @fabianofranz

bgrant0607 commented:
cc @RichieEscarez

kargakis commented on pkg/kubectl/cmd/swagger_doc.go:
```diff
@@ -0,0 +1,65 @@
+	for _, model := range args {
+		if cmdutil.GetFlagBool(cmd, "search") {
+			kubectl.SearchForModel(model, apiV, out, client)
+		} else {
```
Make `GetModelDescription` and `SearchForModel` return errors instead of `os.Exit` so they can be used directly in return statements (will make this else clause redundant).

andronat replied:
Great, thanks! I am on it.

nikhiljindal commented:
Thanks a lot @andronat. This is exciting stuff!
I recommend reading https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md for our general kubectl conventions.

For ex the current proposed ```kubectl doc Pod``` doesnt follow our ```kubectl VERB NOUN``` convention.
We can instead have something like ```kubectl describe model pods``` for returning pods model. 
```kubectl describe model``` can return a list of all models (pods, replication controllers, nodes, etc).
```kubectl describe operation /api/v1/namespaces/x/pods GET``` can describe the GET operation at that path.
```kubectl describe operation /api/v1/namespaces/x/pods``` can return a list of all valid methods for that path.
I am sure @bgrant0607 and @smarterclayton will have better suggestions.

nikhiljindal commented on docs/user-guide/kubectl/kubectl.md:
```diff
@@ -86,6 +86,7 @@ kubectl
 * [kubectl create](kubectl_create.md)	 - Create a resource by filename or stdin
 * [kubectl delete](kubectl_delete.md)	 - Delete resources by filenames, stdin, resources and names, or by resources and label selector.
 * [kubectl describe](kubectl_describe.md)	 - Show details of a specific resource or group of resources
+* [kubectl doc](kubectl_doc.md)	 - Print model's documentation through swagger's API.
```
I think we can do without specifying "swagger" here. Many kubectl users might not know what swagger is and dont need to.
If we decide to change this to kubectl get kind, then we can say "Show the fields of a given kind"

andronat replied:
done

nikhiljindal commented on pkg/client/unversioned/client.go:
```diff
@@ -171,6 +175,49 @@ func (c *Client) ValidateComponents() (*api.ComponentStatusList, error) {
+
+	// This check also takes care the case that kubectl is newer than the running endpoint
+	if stringDoesntExistIn(version, vers.Versions) {
+		return nil, fmt.Errorf("API version: %s is not supported by the server. Use one of: %v", version, vers.Versions)
```
Will be good to say that the api version can be changed by using the --api-version flag

andronat replied:
We changed this functionality by using a prefix in models and taking the version from the urls for operations.

nikhiljindal replied:
What if users dont specify the apiversion, for ex: ```kubectl doc Pod``` do we use the latest version by default (like print v1.Pod, in this case) or are users required to write ```kubectl doc v1.Pod```
We could also support ```kubectl doc Pod --api-version=v2```, since api-version is expected to work across all commands.

andronat replied:
Yes version is optional, we always look on the latest that exist by default.

I didn't know that `--api-version` is a universal flag, I added in this command myself arbitrary. Do you want me to support it?

JanetKuo replied:
Yes, `--api-version` is a global flag. We should support that.

nikhiljindal commented on pkg/kubectl/swagger_doc.go:
```diff
@@ -0,0 +1,170 @@
+	"os"
+	"strings"
+
+	// "github.com/davecgh/go-spew/spew"
```
remove?

andronat replied:
done

nikhiljindal commented on pkg/kubectl/swagger_doc.go:
```diff
@@ -0,0 +1,170 @@
+	swaggerSchema, err := kubeClient.SwaggerSchema(apiVer)
+	if err != nil {
+		fmt.Printf("Couldn't read swagger schema from server: %v\n", err)
+		os.Exit(1)
```
return err

andronat replied:
Fixed that so all functions return an error

nikhiljindal commented on pkg/kubectl/swagger_doc.go:
```diff
@@ -0,0 +1,170 @@
+
+func printModelInfo(w io.Writer, model *swagger.NamedModel, apiVer string) {
+	trimVer := func(str string) string {
+		return strings.TrimPrefix(str, apiVer)
```
Shouldnt this be TrimPrefix(str, apiVer+".")?

andronat replied:
I have already appended the "." in PrintModelDescription.

nikhiljindal commented on pkg/kubectl/swagger_doc.go:
```diff
@@ -0,0 +1,170 @@
+func similarName(modelName, inModel string) bool {
+	mL := strings.ToLower(modelName)
+	iL := strings.ToLower(inModel)
+	return strings.Contains(mL, iL) || strings.Contains(iL, mL)
```
I think just the first condition should be sufficient. We should return resources that contain the user specified name. Dont need to return resources whose name is contained in the user specified name.

andronat replied:
In general I would like to have that as someone might want to search for similar words and he might looks for a prefix of a word. e.g. User writes `Pod` but he looks for `SpecPod` or any other prefix.

Do you want me to do the change?

nikhiljindal commented on pkg/kubectl/swagger_doc.go:
```diff
@@ -0,0 +1,170 @@
+	if prop.Ref != nil {
+		return *prop.Ref
+	} else if *prop.Type == "array" && prop.Items.Ref != nil {
+		return *prop.Items.Ref
```
```kubectl describe Pod.PodSpec.Containers``` and ```kubectl describe Container``` will give the same output?
I was expecting that Pod.PodSpec.Containers will say that Containers is an array of Container and then output the Container model.

nikhiljindal commented on pkg/kubectl/swagger_doc.go:
```diff
@@ -0,0 +1,170 @@
+	} else if *prop.Type == "array" && prop.Items.Ref != nil {
+		return *prop.Items.Ref
+	}
+	return ""
```
This will give an error when inModel is a primitive type field, for ex: Pod.PodSpec.RestartPolicy.
Is that intended? We could just return the swagger description for that field.

andronat replied:
Yes that was intended as we want to describe only models and the primitive types are always a field in them. Do you want me to change that?

nikhiljindal commented on pkg/kubectl/swagger_doc.go:
```diff
@@ -0,0 +1,170 @@
+	return dotModel[0], fieldsPath
+}
+
+func getNextModel(prop *swagger.ModelProperty) string {
```
Its not clear what do you mean by "Next" model?
I think naming this getModelName(prop) is fine.

andronat replied:
the idea is:

I have a field of a model which points to an other model. This functions is retrieving that pointed model. So i think that getModelName is not sufficient :/

nikhiljindal commented on pkg/kubectl/swagger_doc.go:
```diff
@@ -0,0 +1,170 @@
+	return ""
+}
+
+func findField(model *swagger.NamedModel, sField string) *swagger.ModelProperty {
```
getProperty seems a better name to me, since it returns a ModelProperty

andronat replied:
fixed

ghodss commented:
> For ex the current proposed kubectl doc Pod doesnt follow our kubectl VERB NOUN convention.
> We can instead have something like kubectl describe kind Pod for returning the model for Pod kind. 
> kubectl describe kind can return a list of all kinds (Pod, ReplicationController, Node, etc).
> kubectl describe operation /api/v1/namespaces/x/pods GET can describe the GET operation at that path.
> kubectl describe operation /api/v1/namespaces/x/pods can return a list of all valid methods for that path.
> I am sure @bgrant0607 and @smarterclayton will have better suggestions.

I wonder if in this context, `kubectl doc` could be read as "kubectl document NOUN" which would then fit the convention as well as be short to type. I dislike overriding describe with a (to my eyes) completely unrelated operation to what describe does (effectively a more elaborate `kubectl get`).

andronat commented:
Sorry for my really late response and the fact that I didn't pushed my latest version of the PR.

I just updated everything and rebased. I will take a look on your comments now and I will act accordingly.

andronat commented:
I changed the tool to use: `kubectl doc model MODEL-NAME` and `kubectl doc operation URL [GET|...]`. Please review.

kargakis commented on pkg/kubectl/cmd/swagger_doc.go:
```diff
@@ -0,0 +1,73 @@
+	"k8s.io/kubernetes/pkg/kubectl"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+)
+
```
Can you add an example of the usage here and plug it into cmd.Example?

kargakis replied:
Actually add all the different uses of the command

andronat replied:
Thanks, I am on it

kargakis commented on pkg/kubectl/cmd/swagger_doc.go:
```diff
@@ -0,0 +1,73 @@
+
+	mode := args[0]
+	search := cmdutil.GetFlagBool(cmd, "search")
+	apiV := kubectl.GetAPIVersion(args[1:], mode)
```
There is an `--api-version` flag you inherit from kubectl, use that to get the api version

kargakis replied:
Nevermind, I wasn't familiar what you were doing here

kargakis commented on pkg/kubectl/cmd/swagger_doc.go:
```diff
@@ -0,0 +1,73 @@
+		return kubectl.SearchForURL(args[1], out, swagSchema)
+	} else if mode == "operation" && len(args) == 3 {
+		return kubectl.PrintURLDescription(args[1], args[2], out, swagSchema)
+	} else {
```
Nuke this else

kargakis commented on pkg/kubectl/swagger_doc.go:
```diff
@@ -0,0 +1,277 @@
+func GetSwaggerSchema(apiVer string, kubeClient client.Interface) (*swagger.ApiDeclaration, error) {
+	swaggerSchema, err := kubeClient.SwaggerSchema(apiVer)
+	if err != nil {
+		return nil, fmt.Errorf("Couldn't read swagger schema from server: %v\n", err)
```
All errors should be lowercased

kargakis commented on pkg/kubectl/swagger_doc.go:
```diff
@@ -0,0 +1,277 @@
+		}
+	}
+	if pointedModel == nil {
+		return fmt.Errorf("No model found.")
```
lowercase

kargakis commented on pkg/kubectl/swagger_doc.go:
```diff
@@ -0,0 +1,277 @@
+		if prop := getProperty(pointedModel, field); prop != nil {
+			next := getNextModel(prop)
+			if next == "" {
+				return fmt.Errorf("Field: %s can not be expanded further.\n", field)
```
lowercase

kargakis commented on pkg/kubectl/swagger_doc.go:
```diff
@@ -0,0 +1,277 @@
+			}
+			pointedModel = allModels[next]
+		} else {
+			return fmt.Errorf("Field: %s doesn't exist.\n", field)
```
Move this on top so that the else will be removed and lowercase

kargakis commented on pkg/kubectl/swagger_doc.go:
```diff
@@ -0,0 +1,277 @@
+			return nil
+		}
+	}
+	return fmt.Errorf("No URL found.\n")
```
remove \n and lowercase

kargakis replied:
`no URL found`

andronat commented:
@kargakis thanks for the comments I fixed them all.

I got rid of the line wrapping library as it was buggy. I wrote my own implementation specific for our case.

nikhiljindal commented:
I discussed this with @bgrant0607 and here are a few discussion points:

* We need a better verb than doc.
* It will be great if we can make our get resource names work with doc. For ex: ```kubectl doc svc``` should return the Service model.
* We do not want users to use Go Struct names. They should be using json field names. For ex: they shouldnt be allowed to do ```kubectl doc PodSpec``` but they can do ```kubectl doc pods.spec``` (I think since users type Pod in kind already, allowing ````kubectl doc Pod``` in addition to ```kubectl get pods``` might be fine)
* Brian mentioned that there should be a way for users to get even the sub models when they get a parent. For ex: ```kubectl doc Pod``` should return the structure for PodSpec as well. I think what this PR has is fine for default. We can add a flag that users can set to get all submodels as well recursively.
* Primitive field types should work as well. ```kubectl doc Pod.spec.restartPolicy``` can emit the description for that field.

nikhiljindal commented:
Passing this on to @JanetKuo since I will be out for next few days.

bgrant0607 commented:
Re. verb:

I think of this as `help`, but deeper command sequences (e.g., `kubectl help create pod`) are less discoverable. It's also unfortunate that we already used `describe`. `detail` might have been more appropriate.

For this case, `explain` would be more correct English-definition-wise than `document`. `define` could also work, thought it shares the same ambiguity regarding establishing the definition vs. getting the definition as `document`.

As for `doc` vs. `document`: We so far don't have short forms of commands, though we do for resources.

Some users might find `doc` more intuitive, since it's a common term (godoc, etc.), but I like `explain`.

`kubectl explain pod`
`kubectl explain pod.spec`
`kubectl explain pod.spec.restartPolicy`

andronat commented:
In response to @nikhiljindal [list](https://github.com/kubernetes/kubernetes/pull/12677#issuecomment-137200616)

 1. as @bgrant0607 suggested I changed `doc` to explain
 2. I am not sure how can I implement this. I have spotted this list of [abbreviations](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/kubectl.go#L93) do you want me to implement a dictionary like this one that has our words on the one hand and map into models? And what about other models that are not in this list? Should I expose them with they struct names or not expose them at all?
 3. same as above
 4. wouldn't this create too much cluttering? and also in what depth should I stop exposing other structs?
 5. I have the feeling that this breaks the pattern. what i mean is that `kubectl doc Pod.spec.restartPolicy` is included in `kubectl doc Pod.spec` and the point of the `.` is to expand an object to read its fields. primitives do not have a field which means that `.` doesn't have the meaning of expansion.

JanetKuo commented:
@andronat thanks for implementing this. I discussed with @bgrant0607 about the questions you have, and here are some more thoughts about `kubectl explain`. 

The benefit of having `kubectl explain` is that it's like a more detailed help document that facilitates the users to write their yaml/json files. By convention, it should work similar to other kubectl commands, which means it should explain the resources that users need to specify in yaml/json, and then only have the users to think about "field" rather than "model" (since fields are the things they care about in yaml/json). Therefore, in short, we need:
- `kubectl explain <resouce>.<field1>.<field1.1>`: probably we don't need to put "model" in the command and we don't need to support "operation". 
  - We don't need to explain things like PodList anymore, since they only care/know about resource and fields.  
- To support the abbreviation you may use the map that you mentioned to map abbreviation to resources, and then you'll need another map for resources to models. 
- We'll need a flag to support getting submodels recursively. 
- When explaining a resource/field, we don't need to mention anything about model; for example, for `pod` we can print `spec: do kubectl explain pod.spec for more details`. 
  - Since now users are not thinking about models, for primitive type fields we'll still need to print its description.

andronat commented:
Hello @JanetKuo thank you for your time and sorry for my late response. I did many of the changes you asked please take a look.

Here is an example with the recursion ON:
```
kubectl explain -R nodes
DESCRIPTION:
Node is a worker node in Kubernetes, formerly known as minion. Each node will have a unique identifier in the cache (i.e. in etcd).

FIELDS:
   spec	<Object>
       externalID	<string>
       podCIDR	<string>
       providerID	<string>
       unschedulable	<boolean>
   status	<Object>
       addresses	<[]Object>
       capacity	<map[string]string>
       conditions	<[]Object>
       nodeInfo	<Object>
       phase	<string>
   apiVersion	<string>
   kind	<string>
   metadata	<Object>
       resourceVersion	<string>
       selfLink	<string>
       uid	<string>
       annotations	<map[string]string>
       deletionGracePeriodSeconds	<integer>
       deletionTimestamp	<string>
       generateName	<string>
       generation	<integer>
       creationTimestamp	<string>
       labels	<map[string]string>
       name	<string>
       namespace	<string>
```

I kept noting the types of the fields as I believe is crucial for the user as he can understand if he needs a new object or a primitive. What I changed though is that I no longer expose the type of the object, I just report `Object`. Also if we need an array I use `[]Object`. I think this is useful but if you disagree I can delete it.

JanetKuo commented:
@andronat thanks, this looks good! We should avoid using flag shorthands unless necessary (since there are only 52 of them), so please remove the use of `-R` and `-S`. 

@bgrant0607 may have more thoughts on if we're keeping the `explain URL` or not, and the use of `Object` in the output.

andronat commented:
Removed short flags and added `api-version`. Please take a look.

andronat commented:
rebased to latest

JanetKuo commented on pkg/kubectl/cmd/explain.go:
```diff
@@ -0,0 +1,107 @@
+kubectl explain -S /api/v1/namespaces`
+	explainLong = `Documentation of resources and operations.
+
+API version in resources can be defined as a prefix: v1.RESOURCE.
```
Does this mean you can do `kubectl explain v1.pods`? Looks like the version part isn't stripped out in the args.

andronat replied:
I followed your suggestion and switched to `--api-version` only. thanks!

JanetKuo commented on pkg/kubectl/cmd/explain.go:
```diff
@@ -0,0 +1,107 @@
+// NewCmdExplain returns a cobra command for swagger docs
+func NewCmdExplain(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "explain RESOURCE | [GET|POST|PUT|DELETE|PATCH] URL",
```
`explain (RESOURCE | [GET|POST|PUT|DELETE|PATCH] URL)` if we're going to support operation documentation.

andronat replied:
fixed

bgrant0607 commented:
I intended this to be targeted at kubectl users, so to the extent possible all the commands and documentation should be in terms of concepts relevant to kubectl users, not API developers. 

Along those lines, I'd drop the URL/operation support for now.

bgrant0607 commented on pkg/kubectl/explain.go:
```diff
@@ -0,0 +1,394 @@
+	modelMode string = "model"
+)
+
+var resource2Model = map[string]string{
```
Ideally we'd use expandResourceShortcut to map short form to long form, and then map to model:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/kubectl.go#L93

I believe there should be a singularizer somewhere, also, that can be used. @JanetKuo

JanetKuo replied:
Singularizer -https://github.com/kubernetes/kubernetes/blob/master/pkg/api/meta/restmapper.go#L162

bgrant0607 commented:
I like the table-style output in the PR description for Pod.status.conditions. The table should have field name, type, required, and description. Object and []Object seem fine for submodels. We need to make it clear that the way to get those object details is to invoke explain on those fields.

andronat commented:
Thanks @bgrant0607! So if I understand correctly I should:

1) fix the map
2) get rid of URL/operation support
3) get back the tables. (I am not sure though yet how this will work for the recursive format)

Is that correct?

bgrant0607 commented:
The tables aren't strictly required, but the information I listed is.

bgrant0607 commented:
For the nested/recursive format, how do you decide how many levels to show? Always 2? 2 for the top level and 1 for any subobjects below that? All, and you just didn't show them?

I think non-nested/recursive is ok, again, provided that it's obvious how to traverse subobjects via field names.

andronat commented:
For the recursion I just go 1 level deep. I though that we wanted that feature. Do you want me to revert it?

andronat commented:
Made the requested changes please take a look.

JanetKuo commented:
@andronat thanks for the update. I think the map needs to be updated since operations like `kubectl explain pod`, `kubectl explain replicationcontroller` (singularized resource name) don't work. For recursive display if we decide to go one level deep we should say so in the flag description (or perhaps we can use some other flags like --level=2, --depth=3; it's up to you).

andronat commented:
@JanetKuo yea, I fixed the problems

JanetKuo commented:
@andronat thanks, this LGTM now. Please update the description of this PR (the first comment) to reflect the latest change. @bgrant0607 may have more comments.

bgrant0607 commented:
Thanks for all the work on this. We can improve it more later. Please rebase and squash.

andronat commented:
I updated the description and squashed the commits.
